openapi/ws/channels.py

```diff
@@ -0,0 +1,187 @@
+        """
+        channel = self.channel(channel_name)
+        event = channel.register(event, callback)
+        await self.connect()
+        await channel.connect()
+        return channel
+
+    async def unregister(
+            self,
+            channel_name: str,
+            event: str,
+            callback: Callable) -> Channel:
+        """Safely unregister a callback from the list of event
+        callbacks for channel_name
+        """
+        channel = self.channel(channel_name, create=False)
+        if channel:
+            channel.unregister(event, callback)
+            if not channel:
+                await channel.disconnect()
+                self.channels.pop(channel.name)
+        return channel
+
+    async def connect(self, next_time=None):
+        """Connect with broker if possible
+        """
+        if self.status in CAN_CONNECT:
+            self.status = self.statusType.connecting
+            await self._connect(next_time)
+
+    async def publish(self, channel_name, event, data=None):
+        """Publish a new ``event`` on a ``channel_name``
+        """
+        await self.connect()
+        msg = compact(event=event, channel=channel_name, data=data)
+        try:
+            await self.broker.publish(msg, self.prefixed(channel_name))
+        except ConnectionRefusedError:
+            self.connection_error = True
+            logger.critical(
+                '%s cannot publish on "%s" channel - connection error',
+                self,
+                channel_name
+            )
+        else:
+            self._connection_ok()
+        return msg
+
+    def prefixed(self, name):
+        if self.namespace and not name.startswith(self.namespace):
+            name = f'{self.namespace}{name}'
+        return name
+
+    async def close(self):
+        """Close channels and underlying broker handler
+        """
+        self.status = self.statusType.closed
+        if self.broker:
+            await self.broker.close()
+
+    def channel(self, name, create=True):
+        name = name.lower()
+        channel = self.channels.get(name)
+        if channel is None and create:
+            channel = Channel(self, name)
+            self.channels[channel.name] = channel
+        return channel
+
+    def event_pattern(self, event):
+        """Channel pattern for an event name
+        """
+        return redis_to_py_pattern(event or '*')
+
+    # INTERNALS
+
+    async def _subscribe(self, channel_name):
+        """Subscribe to the remote server
+        """
+        await self.broker.bind(self.prefixed(channel_name))
+
+    async def _unsubscribe(self, channel_name):
+        pass
+
+    def _connection_ok(self):
+        if self.connection_error:
+            logger.warning(
+                'connection with %s established - all good',
+                self
+            )
+            self.connection_error = False
+        else:
+            return True
```

Comment:
Why `return True` here? Also, I'm not sure what the `connection_error` property is for. We seem to be going ahead with publishing even if it's set and then clearing it if we succeed.

If we run into an error why not raise and let the caller handle? That way we can avoid failing silently to the caller

Reply:
We also have a `status` which duplicates the placement of state a little here

Reply:
this is the handler and logger